### PR TITLE
[StructuralMechanics][Solver]Duplicated model

### PR DESCRIPTION
**Pull request by loumalouomega**
This error was in the convection-diffusion #2229

### CODE REVIEW

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py [hunk 1]
```diff
@@ -211,15 +211,15 @@ def ImportModelPart(self):
             self._ImportModelPart(self.main_model_part, self.settings["model_import_settings"])
 
     def PrepareModelPart(self):
+        # This will be removed once the Model is fully supported! => It wont e necessary anymore
+        if not self.model.HasModelPart(self.main_model_part.Name):
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py]: I dont think this can be here
The CheckAndPrepare things are executed afterwards, this means that the submodelparts that are created in there wont be in the first lvl of the model, which is currently still needed!

Why did you move it there?

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py [hunk 2]
```diff
@@ -339,11 +339,8 @@ def import_constitutive_laws(self):
         materials_filename = self.settings["material_import_settings"]["materials_filename"].GetString()
         if (materials_filename != ""):
             import read_materials_process
-            # Create a dictionary of model parts.
-            Model = KratosMultiphysics.Model()
-            Model.AddModelPart(self.main_model_part)
             # Add constitutive laws and material properties from json file to model parts.
-            read_materials_process.ReadMaterialsProcess(Model, self.settings["material_import_settings"])
+            read_materials_process.ReadMaterialsProcess(self.model, self.settings["material_import_settings"])
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:343]: 👍

**Review by philbucher — CHANGES_REQUESTED** [review]
Thx but I have one concern

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py (hunk 1 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py]: In that case should be moved before assigning the materials

**Comment by philbucher** [comment]
Now you add the modelpart twice to the model

**Comment by loumalouomega** [comment]
No, because you check if the model has the modelpart:

~~~py
if not self.model.HasModelPart(self.main_model_part.Name):
~~~

**Comment by loumalouomega** [comment]
it is added just once

**Comment by philbucher** [comment]
Inhabe to go, will take a look first thing in the morning

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py [hunk 3]
```diff
@@ -219,7 +219,7 @@ def PrepareModelPart(self):
         # This will be removed once the Model is fully supported! => It wont e necessary anymore
         if not self.model.HasModelPart(self.main_model_part.Name):
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:220]: what I don't get is why you want to add the modelpart twice to the model
Here ...

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py [hunk 4]
```diff
@@ -434,6 +431,10 @@ def _execute_after_reading(self):
         import check_and_prepare_model_process_structural
         check_and_prepare_model_process_structural.CheckAndPrepareModelProcess(self.main_model_part, params).Execute()
 
+        # This will be removed once the Model is fully supported! => It wont e necessary anymore
+        if not self.model.HasModelPart(self.main_model_part.Name):
```
Comment (philbucher) [inline applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py]: ... and here

Can you explain pls?

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py (hunk 3 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:220]: It is added just once, because the check is done "if not self.model.HasModelPart(self.main_model_part.Name):" before adding it

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py:220]: Here is added in case that the problem uses restart

File: applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py (hunk 4 above)
Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py]: Here is added in the normal case that the restart is not used. Besides it is added just once, because the check is done before computing

Comment (philbucher) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py]: Ok now I finally understood
this is needed for the ReadMaterialProcess right?

Can you please add a comment abt this?

Comment (loumalouomega) [reply applications/StructuralMechanicsApplication/python_scripts/structural_mechanics_solver.py]: Yes

**Review by philbucher — APPROVED** [review]
thx
